File: GenAI-dev/10-MCP/host.py
"""
Host MCP + Azure OpenAI.

Flujo:
  1. Lanza server.py como subproceso MCP (stdio) y abre una ClientSession.
  2. list_tools() → traduce los tools MCP al formato de function-calling de OpenAI.
  3. El LLM (Azure OpenAI) decide qué tool llamar; el host ejecuta la llamada
     vía session.call_tool() y le devuelve el resultado al modelo (bucle ReAct).
  4. Cuando el modelo deja de pedir tools, imprime la respuesta final.

Uso:
    python host.py "¿qué hora es en Tokyo y cuánto es 23*7+4?"
    python host.py                # REPL interactivo
"""

import asyncio
import json
import os
import sys

from dotenv import load_dotenv
from mcp import ClientSession
from mcp.client.stdio import StdioServerParameters, stdio_client
from openai import AsyncAzureOpenAI

load_dotenv()

SYSTEM = (
    "Eres un asistente útil y conciso. Dispones de herramientas vía MCP. "
    "Úsalas cuando aporten datos exactos (cálculos, hora, notas). "
    "Responde en el idioma de la pregunta."
)

# Lanza el servidor MCP con el MISMO intérprete del venv.
SERVER = StdioServerParameters(
    command=sys.executable,
    args=[os.path.join(os.path.dirname(__file__), "server.py")],
)


def to_openai_tools(mcp_tools) -> list[dict]:
    """Traduce tools MCP → esquema function-calling de OpenAI."""
    return [
        {
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description or "",
                "parameters": t.inputSchema,
            },
        }
        for t in mcp_tools
    ]


def tool_result_text(result) -> str:
    """Concatena el texto de un CallToolResult."""
    parts = [c.text for c in result.content if getattr(c, "type", None) == "text"]
    return "\n".join(parts) if parts else "(sin contenido)"


async def chat(session: ClientSession, llm: AsyncAzureOpenAI, deployment: str, question: str) -> str:
    tools = to_openai_tools((await session.list_tools()).tools)
    messages = [
        {"role": "system", "content": SYSTEM},
        {"role": "user", "content": question},
    ]

    while True:
        resp = await llm.chat.completions.create(
            model=deployment, messages=messages, tools=tools, temperature=0
        )
        msg = resp.choices[0].message

        if not msg.tool_calls:
            return msg.content or ""

        # Registra la decisión del modelo (asistente con tool_calls)
        messages.append({
            "role": "assistant",
            "content": msg.content,
            "tool_calls": [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {"name": tc.function.name, "arguments": tc.function.arguments},
                }
                for tc in msg.tool_calls
            ],
        })

        # Ejecuta cada tool vía MCP y devuelve el resultado al modelo
        for tc in msg.tool_calls:
            args = json.loads(tc.function.arguments or "{}")
            print(f"  [tool] {tc.function.name}({args})")
            result = await session.call_tool(tc.function.name, args)
            messages.append({
                "role": "tool",
                "tool_call_id": tc.id,
                "content": tool_result_text(result),
            })


async def main():
    deployment = os.getenv("AZURE_OPENAI_DEPLOYMENT")
    llm = AsyncAzureOpenAI(
        api_key=os.getenv("AZURE_OPENAI_API_KEY"),
        azure_endpoint=os.getenv("AZURE_OPENAI_ENDPOINT"),
        api_version=os.getenv("AZURE_OPENAI_API_VERSION", "2025-01-01-preview"),
    )

    async with stdio_client(SERVER) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            tools = (await session.list_tools()).tools
            print(f"[mcp] conectado. tools: {[t.name for t in tools]}\n")

            cli = " ".join(sys.argv[1:]).strip()
            if cli:
                print(await chat(session, llm, deployment, cli))
                return

            print("Host MCP — escribe tu pregunta (Ctrl-C para salir)\n")
            try:
                while True:
                    q = input("> ").strip()
                    if q:
                        print("\n" + await chat(session, llm, deployment, q) + "\n")
            except (KeyboardInterrupt, EOFError):
                print("\nbye")


if __name__ == "__main__":
    asyncio.run(main())

File: GenAI-dev/10-MCP/client-ts/src/host.ts
/**
 * Host MCP + Azure OpenAI — equivalente TypeScript de host.py.
 *
 * Flujo:
 *   1. Conecta al servidor MCP remoto vía client.ts (Streamable HTTP / SSE).
 *   2. listTools() → traduce los tools MCP al formato function-calling de OpenAI.
 *   3. El LLM (Azure OpenAI) decide qué tool llamar; el host ejecuta la llamada
 *      vía callTool() y devuelve el resultado al modelo (bucle ReAct).
 *   4. Cuando el modelo deja de pedir tools, imprime la respuesta final.
 *
 * Uso:
 *   node dist/index.js "¿qué hora es en Tokyo y cuánto es 23*7+4?"
 *   node dist/index.js                # REPL interactivo
 */

import { AzureOpenAI } from "openai";
import type {
  ChatCompletionMessageParam,
  ChatCompletionTool,
} from "openai/resources/index.js";
import { McpClientWrapper, McpToolInfo } from "./client.js";
import { createInterface } from "readline";

const SYSTEM_PROMPT =
  "Eres un asistente útil y conciso. Dispones de herramientas vía MCP. " +
  "Úsalas cuando aporten datos exactos (cálculos, hora, notas). " +
  "Responde en el idioma de la pregunta.";

/**
 * Traduce tools MCP → esquema function-calling de OpenAI.
 */
function toOpenAITools(mcpTools: McpToolInfo[]): ChatCompletionTool[] {
  return mcpTools.map((t) => ({
    type: "function" as const,
    function: {
      name: t.name,
      description: t.description,
      parameters: t.inputSchema,
    },
  }));
}

/**
 * Bucle ReAct: LLM elige tool → callTool(MCP) → resultado → LLM.
 * Repite hasta que el modelo devuelve una respuesta sin tool_calls.
 */
export async function chat(
  mcpClient: McpClientWrapper,
  llm: AzureOpenAI,
  deployment: string,
  question: string,
  tools: ChatCompletionTool[]
): Promise<string> {
  const messages: ChatCompletionMessageParam[] = [
    { role: "system", content: SYSTEM_PROMPT },
    { role: "user", content: question },
  ];

  while (true) {
    const resp = await llm.chat.completions.create({
      model: deployment,
      messages,
      tools,
      temperature: 0,
    });

    const msg = resp.choices[0].message;

    // Sin tool_calls → respuesta final
    if (!msg.tool_calls || msg.tool_calls.length === 0) {
      return msg.content || "";
    }

    // Registrar la decisión del modelo
    messages.push({
      role: "assistant",
      content: msg.content,
      tool_calls: msg.tool_calls.map((tc) => ({
        id: tc.id,
        type: "function" as const,
        function: {
          name: tc.function.name,
          arguments: tc.function.arguments,
        },
      })),
    });

    // Ejecutar cada tool vía MCP
    for (const tc of msg.tool_calls) {
      const args = JSON.parse(tc.function.arguments || "{}");
      console.log(`  [tool] ${tc.function.name}(${JSON.stringify(args)})`);

      const result = await mcpClient.callTool(tc.function.name, args);
      messages.push({
        role: "tool",
        tool_call_id: tc.id,
        content: result,
      });
    }
  }
}

/**
 * Punto de entrada principal.
 */
export async function main(): Promise<void> {
  const serverUrl = process.env.MCP_SERVER_URL;
  if (!serverUrl) {
    console.error("Error: MCP_SERVER_URL no está definida en .env");
    process.exit(1);
  }

  const deployment = process.env.AZURE_OPENAI_DEPLOYMENT;
  if (!deployment) {
    console.error("Error: AZURE_OPENAI_DEPLOYMENT no está definida en .env");
    process.exit(1);
  }

  // Conectar al servidor MCP
  const mcpClient = await McpClientWrapper.connect(serverUrl);

  try {
    // Listar tools disponibles
    const mcpTools = await mcpClient.listTools();
    console.log(
      `[mcp] tools: [${mcpTools.map((t) => t.name).join(", ")}]\n`
    );

    const openaiTools = toOpenAITools(mcpTools);

    // Crear cliente Azure OpenAI
    const llm = new AzureOpenAI({
      apiKey: process.env.AZURE_OPENAI_API_KEY,
      endpoint: process.env.AZURE_OPENAI_ENDPOINT,
      apiVersion:
        process.env.AZURE_OPENAI_API_VERSION || "2025-01-01-preview",
    });

    // Modo CLI (argumento) o REPL interactivo
    const cliQuestion = process.argv.slice(2).join(" ").trim();

    if (cliQuestion) {
      const answer = await chat(
        mcpClient,
        llm,
        deployment,
        cliQuestion,
        openaiTools
      );
      console.log(answer);
    } else {
      console.log("Host MCP — escribe tu pregunta (Ctrl-C para salir)\n");

      const rl = createInterface({
        input: process.stdin,
        output: process.stdout,
      });

      const askQuestion = (): void => {
        rl.question("> ", async (q: string) => {
          const trimmed = q.trim();
          if (trimmed) {
            try {
              const answer = await chat(
                mcpClient,
                llm,
                deployment,
                trimmed,
                openaiTools
              );
              console.log("\n" + answer + "\n");
            } catch (err) {
              console.error(
                `Error: ${err instanceof Error ? err.message : String(err)}\n`
              );
            }
          }
          askQuestion();
        });
      };

      rl.on("close", () => {
        console.log("\nbye");
      });

      askQuestion();

      // Esperar a que el REPL cierre antes de desconectar
      await new Promise<void>((resolve) => {
        rl.on("close", resolve);
      });
    }
  } finally {
    await mcpClient.close();
  }
}
